### PR TITLE
Refactor list/dict/set comprehensions and literal syntax

### PR DESCRIPTION
**Pull request by chris-rands**
This pull request addresses #2085

- [X] I hereby agree to dual licence this and any previous contributions under both
the _Biopython License Agreement_ **AND** the _BSD 3-Clause License_.

- [X] I have read the ``CONTRIBUTING.rst`` file, have run ``flake8`` locally, and
understand that AppVeyor and TravisCI will be used to confirm the Biopython unit
tests and style checks pass with these changes.

- [X] I have added my name to the alphabetical contributors listings in the files
``NEWS.rst`` and ``CONTRIB.rst`` as part of this pull request, am listed
already, or do not wish to be listed. (*This acknowledgement is optional.*)

### CODE REVIEW

File: Bio/Data/CodonTable.py [hunk 1]
```diff
@@ -274,18 +274,18 @@ def list_ambiguous_codons(codons, ambiguous_nucleotide_values):
...
-                            for codon in codons).issuperset(set(meanings)))
-    c3_list = sorted(letter for (letter, meanings)
+                     if {codon[1]
+                         for codon in codons}.issuperset(set(meanings)))
+    c3_list = sorted(letter for letter, meanings
                      in ambiguous_nucleotide_values.items()
-                     if set(codon[2]
-                            for codon in codons).issuperset(set(meanings)))
+                     if {codon[2]
+                         for codon in codons}.issuperset(set(meanings)))
```
Comment (peterjc) [inline Bio/Data/CodonTable.py]: Put this and the two cases just above on one line now?

**Comment by peterjc** [comment]
I was wondering if you found a tool or wrote a script to do this, but looking at the commit log I'm guessing this was mostly by hand? If so, you have a lot of patience - and we should merge this soon to avoid conflicts if we delay it.

File: Bio/Data/CodonTable.py (hunk 1 above)
Comment (peterjc) [reply Bio/Data/CodonTable.py]: Would put it just over 80, but would I think be more legible. Hmm, up to you - judgement time.

**Comment by peterjc** [comment]
Flake8 failed again, a bit cryptic:

```
pydocstyle.parser.ParseError: Cannot parse file.
```

Running locally single threaded may be more informative, but I wonder if my plugin ``flake8-rst-docstrings`` is at fault?

**Comment by peterjc** [comment]
Flake8 failed again (but only under python 2.7), a bit cryptic on the TravisCI logs:

```
pydocstyle.parser.ParseError: Cannot parse file.
```

I'll try to reproduce the Python 2.7 flake8 issue as it might be my ``flake8-rst-docstrings`` plugin at fault - it doesn't crash locally with flake under Python 3.7.

File: Bio/Data/CodonTable.py (hunk 1 above)
Comment (chris-rands) [reply Bio/Data/CodonTable.py]: It does take the lines >80, i'll leave it as it is unless you feel strongly

**Comment by chris-rands** [comment]
I did this mostly manually, i underestimated the task actually, if i did this again i would try to find another way

merging soon to avoid conflicts sounds good- let me know if i should do anything else. I'm not sure on the flake8 fail either

**Comment by peterjc** [comment]
Not seeing the failure locally, re-started that TravisCI job on the odd chance the problem goes away.

**Comment by chris-rands** [comment]
Just added the last fixes (I mistakenly thought these couldn't be changed written before). Now the whole code-base should be clean (`flake8 --isolated --select C4 biopython/` does not find anything)

Guess we could consider adding the flake8-comprehensions plugin to the Biopython `.flake8` to check future PRs?

TravisCI still failed on the re-boot, google didn't reveal much on this for me, just 1 possibly relevant [link](https://github.com/AtomLinter/linter-flake8/issues/670)

**Comment by peterjc** [comment]
From a recent TravisCI on ``master`` which passed, https://travis-ci.org/biopython/biopython/jobs/538814586

```
style create: /home/travis/build/biopython/biopython/.tox/style
style installdeps: flake8, flake8-docstrings, flake8-blind-except, flake8-rst-docstrings, flake8-bugbear;python_version>="3.5", restructuredtext_lint, doc8
style installed: chardet==3.0.4,configparser==3.7.4,doc8==0.8.0,docutils==0.14,entrypoints==0.3,enum34==1.1.6,flake8==3.7.7,flake8-blind-except==0.1.1,flake8-docstrings==1.3.0,flake8-polyfill==1.0.2,flake8-rst-docstrings==0.0.9,functools32==3.2.3.post2,mccabe==0.6.1,pbr==5.2.0,pycodestyle==2.5.0,pydocstyle==3.0.0,pyflakes==2.1.1,restructuredtext-lint==1.3.0,six==1.12.0,snowballstemmer==1.2.1,stevedore==1.30.1,typing==3.6.6
```

From this failing PR,

```
style create: /home/travis/build/biopython/biopython/.tox/style
style installdeps: flake8, flake8-docstrings, flake8-blind-except, flake8-rst-docstrings, flake8-bugbear;python_version>="3.5", restructuredtext_lint, doc8
style installed: chardet==3.0.4,configparser==3.7.4,doc8==0.8.0,docutils==0.14,entrypoints==0.3,enum34==1.1.6,flake8==3.7.7,flake8-blind-except==0.1.1,flake8-docstrings==1.3.0,flake8-polyfill==1.0.2,flake8-rst-docstrings==0.0.9,functools32==3.2.3.post2,mccabe==0.6.1,pbr==5.2.0,pycodestyle==2.5.0,pydocstyle==3.0.0,pyflakes==2.1.1,restructuredtext-lint==1.3.0,six==1.12.0,snowballstemmer==1.2.1,stevedore==1.30.1,typing==3.6.6
```

i.e. Same versions, so it probably isn't bad luck timing with a recent release breaking things.

**Comment by peterjc** [comment]
I've just asked TravisCI to run the API building step, if there is an RST problem (as the error hints) that might throw light on it.

I want to merge this, but not it if is going to break TravisCI on the master branch. I might try splitting up the changes into batches (e.g. ``Tests/``, ``Scripts/``, ...) and applying them one by one - which tell us if it is just some of the files triggering it (and if it doesn't work that's interesting too).

**Comment by peterjc** [comment]
The API docs built fine with Sphinx. Hmm.

**Comment by chris-rands** [comment]
Good news- I think my last commit fixed the TravisCI failure, so it's the changes in `Doc/` (most likely I guess) or `Scripts/` that were the problem apparently

**Comment by peterjc** [comment]
That is extremely strange - but let's merge this (once the automated tests have finished).

Then could you make a new pull request(s) for ``Scripts/``, then ``Doc/``, and then all being well, adding the ``flake8-comprehensions`` plugin to ``.travis-tox.ini`` and ``CONTRIBUTING.rst`` instructions?

**Comment by chris-rands** [comment]
Thanks Peter, yes, sounds like a plan

**Comment by peterjc** [comment]
Thank you, part one merged.

**Pull request merged**